wordpress-mobile/WordPress-Android
# My Site Dashboard: Phase 2: Address PTR Crash

Fixes #15640 

This PR address the following issues with PTR in the My Site View
(1) Refactors `DomainRegistrationSource`:
- Use a mutableMap of  CancellableContinuation instead of a single continuation. 
- Added a check for duplicate requests for the same site id
- Remove/cancel continuations during OnPlansFetched
(2) Removes `distinctUntilChanged()` when the feature flag is on. Note: This _may_ be a temporary solution and as such, the fun was added to both `MySiteViewModel` and `SiteSourceManager`. A decision will be made shortly on whether this will remain, but for now it solves the issue of partial state updates not making it back to the viewModel and therefore prohibiting the refresh indicator from being removed.

@startuptester - if you have time, do your magic and see if this solves what you were seeing. Thank  you.

This PR goes round trip to the API level, so you will see valid Post cards.  
cc @osullivanchris 

**To test:**
Prerequisites
- Launch the app
- Navigate to My Site -> App Settings -> Debug Settings
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on
- Restart the app if needed

#### Test 1 - Multiple PTR requests
-  Launch the app
- Navigate to the My Site tab
Repeat the following scenario multiple times
- Swipe down on the view to trigger the refresh
- Note: Refresh indicator is shown and is then removed from the view

#### Test 1 - Switching sites
-  Launch the app
- Navigate to the My Site tab
- Swipe down on the view to trigger the refresh
- Note: Refresh indicator is shown and is then removed from the view
Repeat the following for each type of  site switch ( non-free site to free site, free site to non-free site, non-free site to non-free site, free site to free site )
- Using the site switcher, select a different site 
- Swipe down on the My Site tab view to trigger the refresh 
- Note: Refresh indicator is shown and is then removed from the view


## Regression Notes
1. Potential unintended areas of impact
My site view shows indefinite progress indicator

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and unit testing

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:
